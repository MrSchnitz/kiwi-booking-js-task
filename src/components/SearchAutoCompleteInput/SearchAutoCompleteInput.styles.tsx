import styled from "styled-components";

export const AutoCompleteWrapper = styled.div`
  position: relative;
`;

export const AutoCompletePopover = styled.div`
  position: absolute;
  top: 50px;
  left: 0;
  z-index: 10;

  border: 1px solid darkgray;
  border-radius: 5px;
  width: 100%;

  background-color: white;

  overflow: hidden;
`;
