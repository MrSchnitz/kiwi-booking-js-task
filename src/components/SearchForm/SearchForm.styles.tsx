import styled from "styled-components";

export const SearchFormWrapper = styled.div`
  padding: 24px;
`;

export const SearchFormPopoverContentWrapper = styled.div`
    max-height: 300px;
    overflow-y: auto;
    
    &::-webkit-scrollbar {
        width: 5px;
    }

    &::-webkit-scrollbar-track {
        background: #ffffff;
    }

    &::-webkit-scrollbar-thumb {
        background: #00A991;
    }

    &::-webkit-scrollbar-thumb:hover {
        background: #007F6D;
    }
`;
