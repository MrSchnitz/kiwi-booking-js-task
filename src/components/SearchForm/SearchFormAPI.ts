import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { all, call, debounce, put, takeEvery } from "redux-saga/effects";
import { RootState } from "../../internals/rootState";
import { SagaIterator } from "redux-saga";
import {
  FlightLocation,
  FlightLocationsResponse,
} from "../../types/FlightLocationTypes";

export enum SearchParam {
  ORIGIN = "origin",
  DESTINATION = "destination",
  DATE_FROM = "dateFrom",
  DATE_TO = "dateTo",
}

export interface ISearchFormData {
  [SearchParam.ORIGIN]: string;
  [SearchParam.DESTINATION]: string;
  [SearchParam.DATE_FROM]: string;
  [SearchParam.DATE_TO]: string;
}

/**
 * API State interface
 */
export interface ISearchFormApi {
  originSearchPhrase: string;
  destinationSearchPhrase: string;
  locations: Array<FlightLocation>;
  searchFormData: ISearchFormData;
}

/**
 * API
 */
class SearchFormApi {
  private static instance: SearchFormApi;

  private constructor() {
    this.handleSearch = this.handleSearch.bind(this);
    this.handleLoadSearchData = this.handleLoadSearchData.bind(this);
    this.saga = this.saga.bind(this);
  }

  public static getInstance(): SearchFormApi {
    if (SearchFormApi.instance) {
      return this.instance;
    }
    this.instance = new SearchFormApi();
    return this.instance;
  }

  /*
   * SLICE
   */
  public getInitialState(): ISearchFormApi {
    return {
      originSearchPhrase: "",
      destinationSearchPhrase: "",
      locations: [],
      searchFormData: {
        origin: "",
        destination: "",
        dateFrom: "",
        dateTo: "",
      },
    };
  }

  public slice = createSlice({
    name: "homePageApiSlice",
    initialState: this.getInitialState(),
    reducers: {
      changeOriginPhrase(state, action: PayloadAction<string>) {
        state.originSearchPhrase = action.payload;
      },
      changeDestinationPhrase(state, action: PayloadAction<string>) {
        state.destinationSearchPhrase = action.payload;
      },
      loadSearchFormData(_, action: PayloadAction<ISearchFormData>) {},
      setLocations(state, action: PayloadAction<Array<FlightLocation>>) {
        state.locations = action.payload;
      },
      setSearchFormData(state, action: PayloadAction<ISearchFormData>) {
        state.searchFormData = action.payload;
      },
    },
  });

  /*
   * SAGAS
   */
  public *handleSearch(action: PayloadAction<string>): SagaIterator {
    const response = yield call(
      fetch,
      `https://api.skypicker.com/locations?term=${action.payload}&location_types=airport`
    );
    const responseBody: FlightLocationsResponse = yield response.json();
    yield put(this.slice.actions.setLocations(responseBody.locations));
  }

  public *handleLoadSearchData({
    payload,
  }: PayloadAction<ISearchFormData>): Generator {
    const searchData: ISearchFormData = { ...payload };
    if (!!searchData[SearchParam.ORIGIN]) {
      const response = yield call(
        this.fetchLocationById,
        searchData[SearchParam.ORIGIN]
      );
      const location: FlightLocation = (response as FlightLocationsResponse)
        .locations[0];
      searchData[SearchParam.ORIGIN] = location.name;
    }
    if (!!searchData[SearchParam.DESTINATION]) {
      const response = yield call(
        this.fetchLocationById,
        searchData[SearchParam.DESTINATION]
      );
      const location: FlightLocation = (response as FlightLocationsResponse)
        .locations[0];
      searchData[SearchParam.DESTINATION] = location.name;
    }

    yield put(this.slice.actions.setSearchFormData(searchData));
  }

  private *fetchLocationById(locationId: string): SagaIterator {
    const response = yield call(
      fetch,
      `https://api.skypicker.com/locations?type=id&id=${locationId}`
    );
    return yield response.json();
  }

  /*
   * SAGA - MAIN
   */
  public *saga(): SagaIterator {
    const { changeOriginPhrase, changeDestinationPhrase, loadSearchFormData } =
      this.slice.actions;
    yield all([
      yield debounce(
        500,
        [changeOriginPhrase.type, changeDestinationPhrase.type],
        this.handleSearch
      ),
      yield takeEvery(loadSearchFormData.type, this.handleLoadSearchData),
    ]);
  }

  /*
   * SELECTORS
   */
  private selectDomain(state: RootState): ISearchFormApi {
    return state.searchFormApiSlice || this.getInitialState();
  }

  public originSearchPhraseSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.originSearchPhrase
  );

  public destinationSearchPhraseSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.destinationSearchPhrase
  );

  public locationsSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.locations
  );

  public searchFormDataSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.searchFormData
  );
}

export default SearchFormApi.getInstance();

export const {
  actions: SearchFormAPI,
  reducer: SearchFormApiReducer,
  name: SearchFormApiName,
} = SearchFormApi.getInstance().slice;

export const {
  originSearchPhraseSelector,
  destinationSearchPhraseSelector,
  locationsSelector,
  searchFormDataSelector,
  saga: SearchFormApiSaga,
} = SearchFormApi.getInstance();
