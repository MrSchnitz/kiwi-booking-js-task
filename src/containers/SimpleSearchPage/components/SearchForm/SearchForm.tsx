import { ChangeEvent, useState } from "react";
import AutoCompleteForm from "../AutoComplete/AutoCompleteForm";

export function SearchForm(props: any) {
  const [originValue, setOriginValue] = useState("");
  const [destinationValue, setDestinationValue] = useState("");
  const [departureValue, setDepartureValue] = useState("");

  const handleChangeDepartureDate = (event: ChangeEvent<any>) => {
    setDepartureValue(event.target.value);
  };

  const handleSearch = () => {
    props.onSearch({
      origin: originValue,
      destination: destinationValue,
      departure: departureValue,
    });
  };

  const handleSelectOrigin = (item) => {
    setOriginValue(item);
  };

  const handleSelectDestination = (item) => {
    setDestinationValue(item);
  };

  return (
    <div>
      {originValue ? (
        <div>{originValue}</div>
      ) : (
        <AutoCompleteForm onSelect={handleSelectOrigin} />
      )}
      <AutoCompleteForm onSelect={handleSelectDestination} />
      <input
        type="date"
        onChange={handleChangeDepartureDate}
        value={departureValue}
      />
      <button onClick={handleSearch} disabled={props.isDisabled}>
        Search
      </button>
    </div>
  );
}
