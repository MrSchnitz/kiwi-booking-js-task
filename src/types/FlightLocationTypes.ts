export interface FlightLocationsResponse {
  locations: FlightLocation[];
}

export interface FlightLocation {
  id: string;
  int_id: number;
  airport_int_id: number;
  active: boolean;
  code: string;
  icao: string;
  name: string;
  slug: string;
  slug_en: string;
  alternative_names: string[];
  rank: number;
  global_rank_dst: number;
  dst_popularity_score: number;
  timezone: string;
  city: City;
  location: LocationCoordinates;
  alternative_departure_points: AlternativeDeparturePoint[];
  tags: Tag[];
  providers: number[];
  special: any[];
  tourist_region: any[];
  car_rentals: any[];
  new_ground: boolean;
  routing_priority: number;
  type: string;
}

export interface AlternativeDeparturePoint {
  id: string;
  distance: number;
  duration: number;
}

export interface City {
  id: string;
  name: string;
  code: string;
  slug: string;
  region: Continent;
  autonomous_territory: null;
  country: Continent;
  continent: Continent;
  subdivision: Continent;
  nearby_country: null;
}

export interface Continent {
  id: string;
  name: string;
  slug: string;
  code?: string;
}

export interface LocationCoordinates {
  lat: number;
  lon: number;
}

export interface Tag {
  tag: string;
  month_to: number;
  month_from: number;
}
