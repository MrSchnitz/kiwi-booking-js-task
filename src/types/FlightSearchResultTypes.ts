export interface FlightSearchResult {
  id: string;
  flyFrom: string;
  flyTo: string;
  cityFrom: string;
  cityCodeFrom: string;
  cityTo: string;
  cityCodeTo: string;
  countryFrom: Country;
  countryTo: Country;
  dTime: number;
  dTimeUTC: number;
  aTime: number;
  aTimeUTC: number;
  nightsInDest: number;
  quality: number;
  distance: number;
  duration: Duration;
  fly_duration: string;
  return_duration: string;
  price: number;
  conversion: Conversion;
  bags_price: BagsPrice;
  baglimit: Baglimit;
  availability: Availability;
  airlines: string[];
  route: Route[];
  booking_token: string;
  deep_link: string;
  tracking_pixel: string;
  facilitated_booking_available: boolean;
  pnr_count: number;
  has_airport_change: boolean;
  technical_stops: number;
  throw_away_ticketing: boolean;
  hidden_city_ticketing: boolean;
  virtual_interlining: boolean;
  mapIdfrom: string;
  mapIdto: string;
  hashtags: string[];
}

export interface Availability {
  seats: number;
}

export interface Baglimit {
  hold_dimensions_sum: number;
  hold_height: number;
  hold_length: number;
  hold_weight: number;
  hold_width: number;
  personal_item_height: number;
  personal_item_length: number;
  personal_item_weight: number;
  personal_item_width: number;
}

export interface BagsPrice {
  "1": number;
  "2": number;
  personal_item: number;
}

export interface Conversion {
  EUR: number;
}

export interface Country {
  code: string;
  name: string;
}

export interface Duration {
  departure: number;
  return: number;
  total: number;
}

export interface Route {
  id: string;
  combination_id: string;
  flyFrom: string;
  flyTo: string;
  cityFrom: string;
  cityCodeFrom: string;
  cityTo: string;
  cityCodeTo: string;
  dTime: number;
  dTimeUTC: number;
  aTime: number;
  aTimeUTC: number;
  airline: string;
  flight_no: number;
  operating_carrier: string;
  operating_flight_no: string;
  fare_basis: string;
  fare_category: string;
  fare_classes: string;
  fare_family: string;
  return: number;
  latFrom: number;
  lngFrom: number;
  latTo: number;
  lngTo: number;
  mapIdfrom: string;
  mapIdto: string;
  bags_recheck_required: boolean;
  vi_connection: boolean;
  guarantee: boolean;
  equipment: null;
  vehicle_type: string;
  following_airport_change?: boolean;
  original_return: number;
}
