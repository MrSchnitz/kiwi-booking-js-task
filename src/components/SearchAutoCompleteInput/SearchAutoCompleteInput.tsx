import React, {
  ChangeEvent,
  SyntheticEvent,
  useEffect,
  useRef,
  useState,
} from "react";
import {
  Button,
  InputField,
  Tile,
  TileGroup,
} from "@kiwicom/orbit-components";
import { ISearchFormData, SearchParam } from "../SearchForm/SearchFormAPI";
import { SearchFormPopoverContentWrapper } from "../SearchForm/SearchForm.styles";
import { CloseCircle } from "@kiwicom/orbit-components/icons";
import { useSearchParams } from "react-router-dom";
import { City, FlightLocation } from "../../types/FlightLocationTypes";
import {
  AutoCompletePopover,
  AutoCompleteWrapper,
} from "./SearchAutoCompleteInput.styles";

interface IProps {
  activeInput: null | SearchParam;
  locations: Array<FlightLocation>;
  searchFormData: ISearchFormData;
  type: SearchParam.ORIGIN | SearchParam.DESTINATION;
  value: string;
  onLocationInputChange(value: string, param: SearchParam): void;
  onSelectLocation(city: City): void;
  onSelectedCityButtonClick(param: SearchParam): void;
}

const SearchAutoCompleteInput = React.memo(
  ({
    activeInput,
    locations,
    searchFormData,
    type,
    value,
    onLocationInputChange,
    onSelectLocation,
    onSelectedCityButtonClick,
  }: IProps) => {
    const wrapperRef = useRef<HTMLDivElement>(null);
    const [isPopoverOpened, setPopoverOpen] = useState<boolean>(false);
    const [searchParams] = useSearchParams();

    useEffect(() => {
      setPopoverOpen(activeInput === type && !!locations.length);
    }, [activeInput, locations.length, type]);

    useEffect(() => {
      document.addEventListener("click", handleDocumentOnClick);

      return () => document.removeEventListener("click", handleDocumentOnClick);
    }, []);

    return (
      <AutoCompleteWrapper ref={wrapperRef}>
        {searchParams.get(type) && searchFormData[type] ? (
          <Button
            type="primary"
            width="100%"
            iconRight={<CloseCircle color="tertiary" />}
            onClick={() => onSelectedCityButtonClick(type)}
          >
            {searchFormData[type]}
          </Button>
        ) : (
          <InputField
            prefix={type === SearchParam.ORIGIN ? "From" : "To"}
            onChange={onInputChange}
            value={value}
            onFocus={onInputFocus}
            onKeyDown={onInputKeyDown}
          />
        )}
        {isPopoverOpened && (
          <AutoCompletePopover>
            <SearchFormPopoverContentWrapper>
              <TileGroup>
                {locations.map((location) => (
                  <Tile
                    key={location.id}
                    onClick={() => onSelectLocation(location.city)}
                  >
                    {location.name}
                  </Tile>
                ))}
              </TileGroup>
            </SearchFormPopoverContentWrapper>
          </AutoCompletePopover>
        )}
      </AutoCompleteWrapper>
    );

    function handleDocumentOnClick(event: MouseEvent): void {
      if (!wrapperRef.current?.contains(event.target as Node)) {
        setPopoverOpen((prevState) => (prevState ? false : prevState));
      }
    }

    function onInputChange(event: SyntheticEvent<HTMLInputElement>): void {
      onLocationInputChange(
        (event as ChangeEvent<HTMLInputElement>).target.value,
        type
      );
    }

    function onInputFocus(): void {
      if (!!value.length) {
        onLocationInputChange(value, type);
        setPopoverOpen(true);
      }
    }

    function onInputKeyDown(event: React.KeyboardEvent | undefined): void {
      if (event?.code === "Escape") {
        setPopoverOpen(false);
      }
    }
  }
);

export default SearchAutoCompleteInput;
