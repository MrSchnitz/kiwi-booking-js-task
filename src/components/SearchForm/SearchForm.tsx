import React, { useCallback, useEffect, useState } from "react";
import {
  createSearchParams,
  useNavigate,
  useSearchParams,
} from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { Button, Heading, Stack } from "@kiwicom/orbit-components";
import {
  destinationSearchPhraseSelector,
  ISearchFormData,
  locationsSelector,
  originSearchPhraseSelector,
  SearchFormAPI,
  searchFormDataSelector,
  SearchParam,
} from "./SearchFormAPI";
import { SearchFormWrapper } from "./SearchForm.styles";
import { City, FlightLocation } from "./SearchForm.types";
import SearchAutoCompleteInput from "../SearchAutoCompleteInput/SearchAutoCompleteInput";
import DateInput from "../DateInput/DateInput";

interface IProps {
  onSearch?(searchParams: ISearchFormData): void;
}

const SearchForm = React.memo(({ onSearch }: IProps) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [activeInput, setActiveInput] = useState<null | SearchParam>(null);
  const [searchParams, setSearchParams] = useSearchParams();

  const originSearchPhrase: string = useSelector(originSearchPhraseSelector);
  const destinationSearchPhrase: string = useSelector(
    destinationSearchPhraseSelector
  );
  const locations: Array<FlightLocation> = useSelector(locationsSelector);
  const searchFormData: ISearchFormData = useSelector(searchFormDataSelector);

  const isSearchDisabled =
    Array.from(searchParams.values()).filter(Boolean).length !== 4;

  useEffect(() => {
    if (Array.from(searchParams.values()).filter(Boolean).length) {
      const searchData: ISearchFormData = {
        [SearchParam.ORIGIN]: "",
        [SearchParam.DESTINATION]: "",
        [SearchParam.DATE_FROM]: "",
        [SearchParam.DATE_TO]: "",
      };
      const searchDataKeys = Object.keys(searchData);

      for (const [k, v] of searchParams.entries()) {
        if (searchDataKeys.includes(k) && !!v) {
          searchData[k as SearchParam] = v;
        }
      }

      dispatch(SearchFormAPI.loadSearchFormData(searchData));
    }
  }, [dispatch, searchParams]);

  const onDateInputChange = useCallback(
    (dateValue: string, param: SearchParam): void => {
      setSearchParams({
        ...Object.fromEntries(searchParams.entries()),
        [param]: dateValue,
      });
      dispatch(
        SearchFormAPI.setSearchFormData({
          ...searchFormData,
          [param]: dateValue,
        })
      );
    },
    [dispatch, searchFormData, searchParams, setSearchParams]
  );

  const onLocationInputChange = useCallback(
    (value: string, param: SearchParam): void => {
      console.log("AAA", value, param);
      setActiveInput(param);
      dispatch(
        param === SearchParam.ORIGIN
          ? SearchFormAPI.changeFromPhrase(value)
          : SearchFormAPI.changeToPhrase(value)
      );
    },
    [dispatch]
  );

  const onSelectLocation = useCallback(
    (city: City): void => {
      if (activeInput) {
        dispatch(SearchFormAPI.setLocations([]));
        dispatch(
          SearchFormAPI.setSearchFormData({
            ...searchFormData,
            [activeInput]: city.name,
          })
        );
        setSearchParams({
          ...Object.fromEntries(searchParams.entries()),
          [activeInput]: city.id,
        });
        setActiveInput(null);
      }
    },
    [dispatch, activeInput, searchParams, searchFormData, setSearchParams]
  );

  const onSelectedCityButtonClick = useCallback(
    (param: SearchParam): void => {
      const searchParamsCopy = Object.fromEntries(searchParams.entries());
      delete searchParamsCopy[param];
      setSearchParams(searchParamsCopy);
    },
    [searchParams, setSearchParams]
  );

  const onSearchClick = (): void => {
    const searchParamsCopy = Object.fromEntries(searchParams.entries());
    const searchData: ISearchFormData = {
      [SearchParam.ORIGIN]: searchParamsCopy[SearchParam.ORIGIN],
      [SearchParam.DESTINATION]: searchParamsCopy[SearchParam.DESTINATION],
      [SearchParam.DATE_FROM]: searchParamsCopy[SearchParam.DATE_FROM],
      [SearchParam.DATE_TO]: searchParamsCopy[SearchParam.DATE_TO],
    };

    if (onSearch) {
      onSearch(searchData);
      return;
    }

    navigate({
      pathname: "search",
      search: `?${createSearchParams({ ...searchData })}`,
    });
  };

  return (
    <SearchFormWrapper>
      <Heading type={"title2"} spaceAfter="medium">
        Search for flights
      </Heading>
      <Stack
        flex={true}
        direction="column"
        justify="center"
        align="center"
        desktop={{ direction: "row" }}
      >
        <Stack>
          <SearchAutoCompleteInput
            type={SearchParam.ORIGIN}
            activeInput={activeInput}
            locations={locations}
            searchFormData={searchFormData}
            value={originSearchPhrase}
            onLocationInputChange={onLocationInputChange}
            onSelectLocation={onSelectLocation}
            onSelectedCityButtonClick={onSelectedCityButtonClick}
          />
        </Stack>
        <Stack>
          <SearchAutoCompleteInput
            type={SearchParam.DESTINATION}
            activeInput={activeInput}
            locations={locations}
            searchFormData={searchFormData}
            value={destinationSearchPhrase}
            onLocationInputChange={onLocationInputChange}
            onSelectLocation={onSelectLocation}
            onSelectedCityButtonClick={onSelectedCityButtonClick}
          />
        </Stack>
        <DateInput
          type={SearchParam.DATE_FROM}
          value={searchFormData[SearchParam.DATE_FROM]}
          prefix="Departure"
          onChange={onDateInputChange}
        />
        <DateInput
          type={SearchParam.DATE_TO}
          value={searchFormData[SearchParam.DATE_TO]}
          prefix="Return"
          onChange={onDateInputChange}
        />
        <Button
          type="primary"
          disabled={isSearchDisabled}
          onClick={onSearchClick}
        >
          Search
        </Button>
      </Stack>
    </SearchFormWrapper>
  );
});

export default SearchForm;
