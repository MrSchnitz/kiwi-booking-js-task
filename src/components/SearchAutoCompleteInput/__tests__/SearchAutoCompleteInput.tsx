import { fireEvent, render } from "@testing-library/react";
import SearchAutoCompleteInput from "../SearchAutoCompleteInput";
import { ISearchFormData, SearchParam } from "../../SearchForm/SearchFormAPI";
import { MemoryRouter } from "react-router-dom";
import { City, FlightLocation } from "../../../types/FlightLocationTypes";

const MOCK_SEARCH_FORM_DATA: ISearchFormData = {
  [SearchParam.ORIGIN]: "Praha",
  [SearchParam.DESTINATION]: "Berlin",
  [SearchParam.DATE_FROM]: "22/07/2022",
  [SearchParam.DATE_TO]: "30/07/2022",
};

const MOCK_EMPTY_SEARCH_FORM_DATA: ISearchFormData = {
  [SearchParam.ORIGIN]: "",
  [SearchParam.DESTINATION]: "",
  [SearchParam.DATE_FROM]: "",
  [SearchParam.DATE_TO]: "",
};

const MOCK_LOCATIONS: Array<Partial<FlightLocation>> = [
  {
    id: "prague",
    name: "Praha",
    city: {
      name: "Praha letiste",
    } as City,
  },
  {
    id: "berlin",
    name: "Berlin",
    city: {
      name: "Berlin letiste",
    } as City,
  },
];

describe("SearchAutoCompleteInput", () => {
  it("should render basic component", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchAutoCompleteInput
          activeInput={null}
          locations={[]}
          searchFormData={MOCK_SEARCH_FORM_DATA}
          type={SearchParam.ORIGIN}
          value={""}
          onLocationInputChange={jest.fn()}
          onSelectLocation={jest.fn()}
          onSelectedCityButtonClick={jest.fn()}
        />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render popover when locations are not empty", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchAutoCompleteInput
          activeInput={SearchParam.ORIGIN}
          locations={MOCK_LOCATIONS as Array<FlightLocation>}
          searchFormData={MOCK_SEARCH_FORM_DATA}
          type={SearchParam.ORIGIN}
          value={""}
          onLocationInputChange={jest.fn()}
          onSelectLocation={jest.fn()}
          onSelectedCityButtonClick={jest.fn()}
        />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with input value", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchAutoCompleteInput
          activeInput={null}
          locations={[]}
          searchFormData={MOCK_EMPTY_SEARCH_FORM_DATA}
          type={SearchParam.ORIGIN}
          value={"Praha"}
          onLocationInputChange={jest.fn()}
          onSelectLocation={jest.fn()}
          onSelectedCityButtonClick={jest.fn()}
        />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("should change location input value", () => {
    const MOCK_INPUT_VALUE = "Barcelona";
    const mockLocationInputChange = jest.fn();

    const { container } = render(
      <MemoryRouter>
        <SearchAutoCompleteInput
          activeInput={SearchParam.ORIGIN}
          locations={MOCK_LOCATIONS as Array<FlightLocation>}
          searchFormData={MOCK_EMPTY_SEARCH_FORM_DATA}
          type={SearchParam.ORIGIN}
          value={""}
          onLocationInputChange={mockLocationInputChange}
          onSelectLocation={jest.fn()}
          onSelectedCityButtonClick={jest.fn()}
        />
      </MemoryRouter>
    );

    const locationInput: HTMLInputElement | null = container.querySelector(
      "input:first-of-type"
    );

    if (locationInput) {
      fireEvent.change(locationInput, { target: { value: MOCK_INPUT_VALUE } });
    }

    expect(mockLocationInputChange).toHaveBeenCalledTimes(1);
    expect(mockLocationInputChange).toHaveBeenCalledWith(
      MOCK_INPUT_VALUE,
      SearchParam.ORIGIN
    );
  });

  it("should select location", () => {
    const mockSelectLocation = jest.fn();

    const { container } = render(
      <MemoryRouter>
        <SearchAutoCompleteInput
          activeInput={SearchParam.ORIGIN}
          locations={MOCK_LOCATIONS as Array<FlightLocation>}
          searchFormData={MOCK_EMPTY_SEARCH_FORM_DATA}
          type={SearchParam.ORIGIN}
          value={""}
          onLocationInputChange={jest.fn()}
          onSelectLocation={mockSelectLocation}
          onSelectedCityButtonClick={jest.fn()}
        />
      </MemoryRouter>
    );

    const locationTile: HTMLDivElement | undefined = Array.from(
      container.querySelectorAll("div")
    ).find((el) => el.textContent === MOCK_LOCATIONS[0].name);

    if (locationTile) {
      fireEvent.click(locationTile);
    }

    expect(mockSelectLocation).toHaveBeenCalledTimes(1);
    expect(mockSelectLocation).toHaveBeenCalledWith(MOCK_LOCATIONS[0].city);
  });

  it("should trigger selected city button click", () => {
    const mockSelectedCityButtonClick = jest.fn();

    const { container } = render(
      <MemoryRouter
        initialEntries={[
          {
            search: `${SearchParam.ORIGIN}=${
              MOCK_SEARCH_FORM_DATA[SearchParam.ORIGIN]
            }`,
          },
        ]}
      >
        <SearchAutoCompleteInput
          activeInput={SearchParam.ORIGIN}
          locations={MOCK_LOCATIONS as Array<FlightLocation>}
          searchFormData={MOCK_SEARCH_FORM_DATA}
          type={SearchParam.ORIGIN}
          value={""}
          onLocationInputChange={jest.fn()}
          onSelectLocation={jest.fn()}
          onSelectedCityButtonClick={mockSelectedCityButtonClick}
        />
      </MemoryRouter>
    );

    const locationButton: HTMLButtonElement | null = container.querySelector(
      "button:first-of-type"
    );

    if (locationButton) {
      fireEvent.click(locationButton);
    }

    expect(mockSelectedCityButtonClick).toHaveBeenCalledTimes(1);
    expect(mockSelectedCityButtonClick).toHaveBeenCalledWith(
      SearchParam.ORIGIN
    );
  });
});
