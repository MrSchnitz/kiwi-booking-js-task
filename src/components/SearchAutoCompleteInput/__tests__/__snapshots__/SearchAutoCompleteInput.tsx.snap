// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchAutoCompleteInput should render basic component 1`] = `
<div>
  <div
    class="sc-dkzDqf fNTABv"
  >
    <div
      class="InputField__Field-sc-jjv066-0 koAYkz"
    >
      <div
        class="InputField__InputContainer-sc-jjv066-2 bFmIrl"
      >
        <div
          class="InputField__Prefix-sc-jjv066-4 kUJKxj"
        >
          From
        </div>
        <input
          aria-labelledby="1"
          class="InputField__Input-sc-jjv066-6 kPMZTv"
          data-state="ok"
          id="1"
          type="text"
          value=""
        />
        <div
          class="InputField__FakeInput-sc-jjv066-1 dmsAbX"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchAutoCompleteInput should render popover when locations are not empty 1`] = `
<div>
  <div
    class="sc-dkzDqf fNTABv"
  >
    <div
      class="InputField__Field-sc-jjv066-0 koAYkz"
    >
      <div
        class="InputField__InputContainer-sc-jjv066-2 bFmIrl"
      >
        <div
          class="InputField__Prefix-sc-jjv066-4 kUJKxj"
        >
          From
        </div>
        <input
          aria-labelledby="2"
          class="InputField__Input-sc-jjv066-6 kPMZTv"
          data-state="ok"
          id="2"
          type="text"
          value=""
        />
        <div
          class="InputField__FakeInput-sc-jjv066-1 dmsAbX"
        />
      </div>
    </div>
    <div
      class="sc-hKMtZM hmtMuM"
    >
      <div
        class="sc-gsnTZi cDgEyT"
      >
        <div
          class="TileGroup__StyledTileGroup-sc-tuzvgn-0 jirIHz"
        >
          <div
            class="TileWrapper__StyledTileWrapper-sc-u2u00m-0 gHuiEQ"
            role="button"
            tabindex="0"
          >
            <div
              class="TileContent__StyledTileContent-sc-1kfck64-0 enwTwt"
            >
              Praha
            </div>
          </div>
          <div
            class="TileWrapper__StyledTileWrapper-sc-u2u00m-0 gHuiEQ"
            role="button"
            tabindex="0"
          >
            <div
              class="TileContent__StyledTileContent-sc-1kfck64-0 enwTwt"
            >
              Berlin
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchAutoCompleteInput should render with input value 1`] = `
<div>
  <div
    class="sc-dkzDqf fNTABv"
  >
    <div
      class="InputField__Field-sc-jjv066-0 koAYkz"
    >
      <div
        class="InputField__InputContainer-sc-jjv066-2 bFmIrl"
      >
        <div
          class="InputField__Prefix-sc-jjv066-4 kUJKxj"
        >
          From
        </div>
        <input
          aria-labelledby="4"
          class="InputField__Input-sc-jjv066-6 kPMZTv"
          data-state="ok"
          id="4"
          type="text"
          value="Praha"
        />
        <div
          class="InputField__FakeInput-sc-jjv066-1 dmsAbX"
        />
      </div>
    </div>
  </div>
</div>
`;
